corgidrp/data.py

```diff
@@ -313,6 +321,36 @@ def get_masked_data(self):
         mask = self.dq>0
         return ma.masked_array(self.data, mask=mask)
 
+    def add_error_term(self, error, err_name):
+        """
+        add a layer of a specific uncertainty on the 3-dim error array extension 
+        
+        Args:
+          error (np.array): 2-d error layer
+          err_name (str): name of the uncertainty layer  
+        """
+        if error.ndim != 2 or error.shape != self.data.shape:
+            raise ValueError("we expect a 2-dimensional error layer with dimensions {0}".format(self.data.shape))
+        
+        err = self.err
+        sh = list(err.shape)
+        if err.ndim > 2:
+            sh[0]+=1
+        else:
+            sh.insert(0,2)
+        
+        new_err = np.resize(err, tuple(sh))
+        #append new error as layer on 3D cube
+        new_err[sh[0] - 1,:, :] = error
+
+        comb_err = np.sqrt(new_err[0,:,:]**2 + error**2)
```

Comment:
I've been thinking about this for a bit. There are cases where we're going to divide by something (e.g. a flat field, the K-gain, even the EMGain) with an error on them as well and so when combining things like that we won't be able to manipulate the errors directly like this. How do we want to handle that? Do we need separate functions? How do we track the errors in that case? Thoughts?

Reply:
I suppose one way is to have a "Layer_N Type" (or something smaller) header keyword that clarifies it. and if we ever need to re-build the combined_error then we also read the type.

Reply:
That is a very valuable comment. We left that out for the sake of simplicity. I have to think about it. To add a multiplicative error we would need not only the error of the data (e.g. flat) but also the data itself to calculate the error propagation, so we would need more optional parameters. The other way would be to do this directly  in the e.g. flatfield step and use this function only for additive errors.

Reply:
It sounds like there are possibly a few different things going on:

1. If we change the units of the data (e.g., by a scale factor), the errors should scale too. I think the function that does this scaling should obviously also change the units of the errors. If it is just a single scale factor, the error on the scale factor doesn't change the SNR of the data. It merely changes the "photometric calibration" of the units. I think we should keep the error map to track the SNR of the data. And we can use header keywords, or something else to track errors that need to be accounted for in the photometric calibration. 
2. For the flat field correction, you can imagine each pixel has a different error on the multiplicative factor. However, we can rewrite a multiplicative error as just an error to be added in quadrature (via standard error propagation equations) with no loss in information. Thus, I don't think this requires special handling.
3. For something like gain, the error on the gain affects the photon noise calculation. I would say the error on the photon noise error is probably small enough for us to not keep track of, but it certainly could just be an extra slice in the error map.

Overall, it feels like to me we can handle it in our current framework. Additionally, I think these are probably small things that we shouldn't focus on now. I'd rather cross these bridges in the future, after we get the main features of the DRP complete.

Reply:
There is a good table of error propagation for different functions (the covariances can be neglected)  in:
https://en.wikipedia.org/wiki/Propagation_of_uncertainty
I think I just keep the additive function as is for now and we come to more complex things if necessary

Reply:
Thanks for the conversation on this. I agree that we can make it work in the current framework. Users will have to make sure to treat everything correctly as we go forward within the individual functions. Including making sure to multiply/divide the error terms appropriately and add the error terms. With the current framework, we'd have to keep track of that in them primary header's history only, as the error headers aren't accessed via e.g. the `update_after_processing_step` function. I think that's probably ok though. @JuergenSchreiber, @JuergenSchreiber any last thoughts on that?